src/flamenco/runtime/program/fd_bpf_loader_v3_program.c

```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
+    /* Update the ProgramData account and record the program bits */
+    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L669-L674 */
+    fd_borrowed_account_t * programdata = NULL;
+    err = fd_instr_borrowed_account_modify_idx( instr_ctx, 1UL, programdata_len, &programdata );
+    if( err!=FD_ACC_MGR_SUCCESS ) {
+      FD_LOG_WARNING(( "Borrowed account lookup for programdata account failed" ));
       return err;
-
-    program_acc_loader_state.discriminant = fd_bpf_upgradeable_loader_state_enum_program;
-    fd_memcpy(&program_acc_loader_state.inner.program.programdata_address, programdata_acc, sizeof(fd_pubkey_t));
-
-    write_result = write_bpf_upgradeable_loader_state( &ctx, program_acc, &program_acc_loader_state );
-    if( FD_UNLIKELY( write_result != FD_ACC_MGR_SUCCESS ) ) {
-      FD_LOG_DEBUG(( "failed to write loader state "));
-      return FD_EXECUTOR_INSTR_ERR_GENERIC_ERR;
     }
 
-    err = fd_account_set_executable2( &ctx, program_acc, program_rec->meta, 1 );
-    if (err != 0)
+    fd_bpf_upgradeable_loader_state_t programdata_loader_state = {
+      .discriminant = fd_bpf_upgradeable_loader_state_enum_program_data,
+      .inner.program_data = {
+        .slot                      = clock.slot,
+        .upgrade_authority_address = (fd_pubkey_t *)authority_key,
+      },
+    };
+    err = fd_bpf_loader_v3_program_write_state( instr_ctx, programdata, &programdata_loader_state );
+    if( err!=FD_BINCODE_SUCCESS ) {
+      FD_LOG_WARNING(( "Bpf state write for programdata account failed" ));
       return err;
-
-    (void)clock_acc;
-    (void)rent_acc;
-
-    return FD_EXECUTOR_INSTR_SUCCESS;
-  } else if ( fd_bpf_upgradeable_loader_program_instruction_is_upgrade( &instruction ) ) {
-    if( ctx.instr->acct_cnt < 7 ) {
+    }
+    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L675-L689 */
+    if( PROGRAMDATA_METADATA_SIZE+buffer_data_len>=programdata->meta->dlen ) {
+      FD_LOG_WARNING(( "ProgramData account too small" ));
+      return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
+    }
+    if( buffer_data_offset>=buffer->meta->dlen ) {
+      FD_LOG_WARNING(( "Buffer account too small" ));
+      return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
+    }
+    uchar * dst_slice       = programdata->data + PROGRAMDATA_METADATA_SIZE;
+    ulong dst_slice_len     = buffer_data_len; 
+    const uchar * src_slice = buffer->const_data + buffer_data_offset;
+    fd_memcpy( dst_slice, src_slice, dst_slice_len );
+
+    /* Update the programdata's metadata */
+    programdata->meta->dlen            = programdata_len;
+    programdata->meta->info.executable = 0U;
+    programdata->meta->info.lamports   = fd_rent_exempt_minimum_balance( instr_ctx->slot_ctx, programdata_len );
+    programdata->meta->info.rent_epoch = 0UL;
```

Comment:
Where are these checks in Agave? These don't line up with the source - do you mind adding a permalink 🙏